litebird_sim/beam_convolution.py

```diff
@@ -0,0 +1,359 @@
+            mmax=default_mmax,
+        )
+
+    else:
+        assert convolution_params.lmax - 4 >= convolution_params.mmax, (
+            "Error in the convolution parameters m_max must be ≦ ℓ_max!"
+            "Here ℓ_max={lmax} and m_max={mmax}!"
+        ).format(lmax=convolution_params.lmax, mmax=convolution_params.mmax)
+
+    if convolution_params.lmax != sky_alms_det.lmax:
+        _slm = sky_alms_det.resize_alm(
+            convolution_params.lmax, convolution_params.lmax, inplace=False
+        ).values
+    else:
+        _slm = sky_alms_det.values
+
+    if (convolution_params.lmax != beam_alms_det.lmax) or (
+        convolution_params.mmax != beam_alms_det.mmax
+    ):
+        _blm = beam_alms_det.resize_alm(
+            convolution_params.lmax, convolution_params.mmax, inplace=False
+        ).values
+    else:
+        _blm = beam_alms_det.values
+
+    if hwp_angle is None:
+        # we cannot simulate HWP, so let's use classic 4pi convolution
+        # FIXME!! we need to understand what to do with the interface here
+        if convolution_params.single_precision:
+            _ftype = np.float32
+            _ctype = np.complex64
+            intertype = Interpolator_f
+        else:
+            _ftype = np.float64
+            _ctype = np.complex128
+            intertype = Interpolator
+
+        inter = intertype(
+            sky=_slm.astype(_ctype),
+            beam=_blm.astype(_ctype),
+            separate=False,
+            lmax=convolution_params.lmax,
+            kmax=convolution_params.mmax,
+            epsilon=convolution_params.epsilon,
+            nthreads=nthreads,
+        )
+        tod_det += inter.interpol(pointings_det.astype(_ftype))[0]
+    else:
+        fullconv = MuellerConvolver(
+            slm=_slm,
+            blm=_blm,
+            mueller=mueller_matrix,
+            lmax=convolution_params.lmax,
+            kmax=convolution_params.mmax,
+            single_precision=convolution_params.single_precision,
+            epsilon=convolution_params.epsilon,
+            nthreads=nthreads,
+        )
+        tod_det += fullconv.signal(ptg=pointings_det, alpha=hwp_angle)
+
+
+def add_convolved_sky(
+    tod,
+    pointings,
+    sky_alms: Union[SphericalHarmonics, Dict[str, SphericalHarmonics]],
+    beam_alms: Union[SphericalHarmonics, Dict[str, SphericalHarmonics]],
+    hwp_angle: Union[np.ndarray, None] = None,
+    mueller_hwp: Union[np.ndarray, None] = None,
+    input_sky_names: Union[str, None] = None,
+    input_beam_names: Union[str, None] = None,
+    convolution_params: Optional[BeamConvolutionParameters] = None,
+    input_sky_alms_in_galactic: bool = True,
+    nthreads: int = 0,
+):
+    """ """
+
+    n_detectors = tod.shape[0]
+
+    if type(pointings) is np.ndarray:
+        assert tod.shape == pointings.shape[0:2]
+
+    for detector_idx in range(n_detectors):
+        if type(pointings) is np.ndarray:
+            curr_pointings_det = pointings[detector_idx, :, :]
```

Comment:
Unlike the `else:` branch, this does not initialize `hwp_angle`, which might be `None`

Reply:
This should be ok, if `hwp_angle ` is None `add_convolved_sky_to_one_detector` uses the standard convolution